feat!: Implement the version module (#1640)
celestiaorg/celestia-app: x/version/README.md

			AppVersion: newAppVersion,
		}
		// set the version in the application to ensure that tendermint is
		// passed the correct value upon rebooting
		keeper.versionSetter.SetProtocolVersion(newAppVersion)
	}
	return resp
}
```

## State

The state in the version module is abnormal because it can be modified by the
party building the binary without breaking the app hash. This is because the
state is not stored in the iavl tree, but rather as a simple predefined golang
map. Its important to note that even though this state is not in the state
store, that is can still cause consensus breaking changes. This is because that
state controls at which height the application will increment the app version.

```go
type Keeper struct {
	chainAppVersions map[string]ChainVersionConfig
}
```

## Events

TODO: Add events

### Usage

The standard usage of the version module is to load the app version height
mappings that are embedded into the binary, but it is also possible to load
custom mappings by passing including a `map[string]ChainVersionConfig` in the
app options using the `version.CustomVersionConfigKey` key.

Mappings are defined by a `map[uint64]int64` where the key is the app version

> and the value is the height at which the app version should be incremented. For

Reviewer: > the key is the app version
and the value is the height at which the app version should be incremented

A few questions on the mapping:
1. Should we enforce that keys of this map are monotonically increasing by 1? The map itself isn't ordered but maybe it should panic if there is a gap in app version keys?
2. Is the app version incremented to one above the key at the value?

Reviewer: Disregard, from tests below I think the statement could be updated to

```suggestion
Mappings are defined by a `map[uint64]int64` where the key is the app version
and the value is the first block height for that app version. For
```

example:

```go
vm := map[uint64]int64{

> 		1: 0,

Reviewer: I guess I'm confused on why the app version doesn't start at 0 and then increment to 1 at some future block. Is an example like the following more reflective of what we expect on BSR and mainnet:

```
vm := map[uint64]int64{
		0: 0,
		1: 1000,
}
```

but even this example would increment the app version from `0` to `1` at block height `0` so I don't think it's accurate.

> 		1: 0,

Reviewer: Disregard, resolved by looking at tests below.

		2: 112093,
		3: 300442,
		4: 420420,
}
```

> The application will convert this mapping into a HeightVersionConfig:

Reviewer: ?

```suggestion
The application will convert this mapping into a HeightRange:
```

is set in the header.

```go
func (app *App) EndBlock(req abci.RequestEndBlock) (res abci.ResponseEndBlock) {
	res = app.BaseApp.EndBlock(req)
	ctx := app.GetContextForDeliverTx([]byte{1})
	return appversion.EndBlocker(ctx, app.VersionKeeper, res)
}

// EndBlocker will modify the app version if the current height is equal to
// a predefined height at which the app version should be changed.
func EndBlocker(ctx sdk.Context, keeper Keeper, resp abci.ResponseEndBlock) abci.ResponseEndBlock {
	newAppVersion := keeper.GetVersion(ctx)
	if ctx.BlockHeader().Version.App != newAppVersion {
		resp.ConsensusParamUpdates.Version = &coretypes.VersionParams{
			AppVersion: newAppVersion,
		}
		// set the version in the application to ensure that tendermint is
		// passed the correct value upon rebooting
		keeper.versionSetter.SetProtocolVersion(newAppVersion)
	}
	return resp
}
```

## State

The state in the version module is abnormal because it can be modified by the
party building the binary without breaking the app hash. This is because the
state is not stored in the iavl tree, but rather as a simple predefined golang

> map. It's important to note that even though this state is not in the state

Reviewer 2: Can you add a sentence where this map is persisted if not in the state? (even if that is explained in more detail below)
